hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +48,272 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+                                        << std::endl;
+            }
+        }
+    m_checked_collision_warnings = true;
+    }
+
+void mpcd::CollisionMethod::beginRigidBodyCollision(uint64_t timestep)
+    {
+    unsigned int num_group = m_embed_group->getNumMembers();
+    ArrayHandle<Scalar4> h_initial_vel(m_initial_velocity,
+                                       access_location::host,
+                                       access_mode::overwrite);
+    ArrayHandle<unsigned int> h_embed_group(m_embed_group->getIndexArray(),
+                                            access_location::host,
+                                            access_mode::read);
+    ArrayHandle<Scalar4> h_vel_embed(m_pdata->getVelocities(),
+                                     access_location::host,
+                                     access_mode::read);
+    for (unsigned int idx = 0; idx < num_group; ++idx)
+        {
+        // collect the initial velocities of the embedded particles
+        const unsigned int particle_index = h_embed_group.data[idx];
+        h_initial_vel.data[idx] = h_vel_embed.data[particle_index];
+        }
+    }
+
+void mpcd::CollisionMethod::accumulateRigidBodyMomenta(uint64_t timestep)
+    {
+    unsigned int num_group = m_embed_group->getNumMembers();
+    ArrayHandle<Scalar4> h_initial_vel(m_initial_velocity,
+                                       access_location::host,
+                                       access_mode::read);
+    ArrayHandle<Scalar3> h_linmom_accum(m_linmom_accum,
+                                        access_location::host,
+                                        access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom_accum(m_angmom_accum,
+                                        access_location::host,
+                                        access_mode::readwrite);
+    ArrayHandle<unsigned int> h_embed_group(m_embed_group->getIndexArray(),
+                                            access_location::host,
+                                            access_mode::read);
+    ArrayHandle<Scalar4> h_postype(m_pdata->getPositions(),
+                                   access_location::host,
+                                   access_mode::read);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::read);
+    ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                     access_location::host,
+                                     access_mode::read);
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+    for (unsigned int idx = 0; idx < num_group; ++idx)
```

Comment:
Before this loop, either memset `h_linmom_accum` and `h_angmom_accum` to zero OR use the new `zeroFill()` method of `GPUArray` before acquiring the array handle.